texk/psutils/psutils-1.21/extractres.1
.TH EXTRACTRES 1 "PSUtils"
.SH NAME
extractres \- filter to extract resources from a PostScript document
.SH SYNOPSIS
.B extractres 
[
.B \-m
] < 
.I document.ps
>
.I output.ps
.SH DESCRIPTION
.I Extractres
extracts resources (fonts, procsets, patterns, files, etc) appearing in a
PostScript document, and puts appropriate
.I %%IncludeResource
comments in the document prologue. The extracted resources are written to
files with the same name as the resource, and an appropriate extension. The
pipeline
.sp
.ce
extractres file.ps | includeres >out.ps
.sp
will move all resources appearing in a document to the document prologue,
removing redundant copies. The output file can then be put through page
re-arrangement filters such as
.I psnup
or 
.I pstops
safely.
.PP
The
.B \-m
option combines the resources of the same name into one file; this must be
used for some programs which download fonts a bit at a time.
.SH AUTHOR
Written by Angus J. C. Duggan.
.SH "SEE ALSO"
psutils(1)
.SH TRADEMARKS
.B PostScript
is a trademark of Adobe Systems Incorporated.
.SH BUGS
.I extractres
does not alter the
.I %%DocumentSuppliedResources
comments.


ChangeLog
2013-10-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* doc: New directory for texlive.info & Co.
	* Makefile.am, configure.ac: Adapted.

2013-10-05  Karl Berry  <karl@tug.org>

	* Build: reorder variables and settings for clarity.
	(TL_CONF_BANNER): restore variable, default is empty string.
	(TL_POSTCONFIGURE): new variable, default true.

2013-09-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop ACLOCAL_AMFLAGS.
	* configure.ac: Use AC_CONFIG_MACRO_DIRS.
	Use KPSE_RECURSE_LIBS in m4/kpse-setup.m4 to recurse into all
	TeX specific libraries.

2013-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-09-01.

	* Makefile.am: Moved definition of cf_silent* from here ...
	* am/recurse.am: ... to here 

2013-08-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add the make target 'texlinks'.

2013-07-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid still more 'parallel make' problems.

2013-07-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid more 'parallel make' problems.

2013-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/*.am: Moved Makefile fragments from libs/am/ and tekk/am/.
	* am/rebuild.am: New Makefile fragment to rebuild a library.
	* am/reconfig.am: New Makefile fragment to reconfigure.
	* am/recurse.am: New Makefile fragment to recurse into subdirs.

	* Makefile.am, configure.ac: Use am/recurse.am.

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Added Automake options 'tar-pax' for paths
	longer than 100 chars and 'no-dist-gzip'.
	* version.ac: Switch 2013 => 2014/dev'
	* tardate.ac: Switch to 2013-06-24.

2013-03-26  Karl Berry  <karl@tug.org>

	* Build: use $MAKE rather than $TL_MAKE for the make clean
	in the unlikely event of a local Makefile.

2013-03-21  Mojca Miklavec  <mojca.miklavec.lists@gmail.com>

	* Build: Remove GNUMAKE variable (after freetype2 as the only library
	which needed it has been fixed).

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-03-01.

2013-02-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Building for Windows does not require 'ln -s'.
	* README.config: Document '--without-ln-s'.

2013-02-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (mandir, infodir): Change texmf => texmf-dist.

2013-01-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-01-24.
	* Makefile.am: Running texlinks in texk/texlive/tl_scripts.

2013-01-18  Karl Berry  <karl@tug.org>

	* Build: remove use of time.

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-11-16.

2012-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-10-24.

2012-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-08-19.

2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_LARGEFILE): switch default to --enable-largefile.
	More info in README.config.

2012-03-12  Karl Berry  <karl@tug.org>

	* README.config: 32-bit rather than 32Bit

2012-03-11  Karl Berry  <karl@tug.org>

	* README.config: text tweaks for poppler/libxpdf/luatex/etc.

2012-01-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report `$enable_largefile'.

2012-01-13  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-01-13.

2011-12-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Replace DISTCHECK_CONFIGURE_FLAGS by
	AM_DISTCHECK_CONFIGURE_FLAGS (for automake-1.11.2).

2011-12-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Use TL specific installation directories only
	for $enable_native_texlive_build=yes (also datarootdir).
	Display the installation directories.

	* README.config, README.distro: Adapted.

2011-10-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-10-05.

2011-08-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-08-10.

2011-07-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am [EXTRA_DIST]: Add m4/ChangeLog.

2011-07-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2011 => 2012/dev.

2011-06-21  Karl Berry  <karl@tug.org>

	* Build: report final exit status.

2011-04-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-06-15.

2011-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am, configure.ac: Conditional 'cross_compiling' =>
	'cross' as in libs/icu, texk/dvipng, and texk/web2c.

	* version.ac: Switch 2011/dev => 2011.

2011-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-04-12.

2011-04-09  Karl Berry  <karl@tug.org>

	* Build (MACOSX_DEPLOYMENT_TARGET): no longer needed per koch.

2011-03-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-03-03.

2011-02-16  Karl Berry  <karl@tug.org>

	* Build: include uname -a in log, mention log if failure.


utils/psutils/psutils-1.17/extractres.pl
@PERL@
# extractres: extract resources from PostScript file
#
# Copyright (C) Angus J. C. Duggan 1991-1995
# See file LICENSE for details.

$0 =~ s=.*/==;
$prog = $0;

%resources = ();		# list of resources included
%merge = ();			# list of resources extracted this time
%extn = ("font", ".pfa", "file", ".ps", "procset", ".ps", # resource extns
	 "pattern", ".pat", "form", ".frm", "encoding", ".enc");
%type = ("%%BeginFile:", "file", "%%BeginProcSet:", "procset",
	 "%%BeginFont:", "font"); # resource types

while (@ARGV) {
   $_ = shift;
   if (/^-m(erge)?$/) { $merge = 1; }
   elsif (/^-/) {
      print STDERR "Usage: $prog [-merge] [file]\n";
      exit 1;
   } else {
      unshift(@ARGV, $_);
      last;
   }
}

if (defined($ENV{TMPDIR})) {	# set body file name
   $body = "$ENV{TMPDIR}/body$$.ps";
} elsif (defined($ENV{TEMP})) {
   $body = "$ENV{TEMP}/body$$.ps";
} elsif (defined($ENV{TMP})) {
   $body = "$ENV{TMP}/body$$.ps";
} else {
   $body = "body$$.ps";
}

open(BODY, $body) && die "Temporary file $body already exists";
open(BODY, ">$body") || die "Can't write file $body";

sub filename {			# make filename for resource in @_
   local($name);
   foreach (@_) {		# sanitise name
      s/[!()\$\#*&\\\|\`\'\"\~\{\}\[\]\<\>\?]//g;
      $name .= $_;
   }
   $name =~ s@.*/@@;		# drop directories
   die "Filename not found for resource ", join(" ", @_), "\n"
      if $name =~ /^$/;
   $name;
}

$output = STDOUT;		# start writing header out
while (<>) {
   if (/^%%BeginResource:/ || /^%%BeginFont:/ || /^%%BeginProcSet:/) {
      local($comment, @res) = split(/\s+/); # look at resource type
      local($type) = defined($type{$comment}) ? $type{$comment} : shift(@res);
      local($name) = &filename(@res, $extn{$type}); # make file name
      $saveout = $output;
      if (!$resources{$name}) {
	 print "%%IncludeResource: $type ", join(" ", @res), "\n";
	 if (!open(RES, $name)) {
	    open(RES, ">$name") || die "Can't write file $name";
	    $resources{$name} = $name;
	    $merge{$name} = $merge;
	    $output = RES;
	 } else {		# resource already exists
	    close(RES);
	    undef $output;
	 }
      } elsif ($merge{$name}) {
	 open(RES, ">>$name") || die "Can't append to file $name";
	 $output = RES;
      } else {			# resource already included
	 undef $output;
      }
   } elsif (/^%%EndResource/ || /^%%EndFont/ || /^%%EndProcSet/) {
      if (defined $output) {
	 print $output $_;
	 close($output);
      }
      $output = $saveout;
      next;
   } elsif ((/^%%EndProlog/ || /^%%BeginSetup/ || /^%%Page:/)) {
      $output = BODY;
   }
   print $output $_
      if defined $output;
}

close(BODY);			# close body output file

open(BODY, $body);		# reopen body for input
while (<BODY>) {		# print it all
   print $_;
}
close(BODY);

unlink($body);			# dispose of body file
@END@
